Swap warning counter to flag in LRScheduler

mlazos opened the pull request:
This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125653
* #125383
* #123753
* #123752
* __->__ #123751

mlazos committed:
Remove out of order warning

[ghstack-poisoned]

janeyx99 commented:
mmmm I think we need a stronger reason to remove this. https://github.com/pytorch/pytorch/pull/21460#issuecomment-499617538

mlazos commented:
> mmmm I think we need a stronger reason to remove this. https://github.com/pytorch/pytorch/pull/21460#issuecomment-499617538

I could change it to a flag possibly. But this is from a BC change five years ago haha I think it's been baked in by this point?

mlazos commented:
> > mmmm I think we need a stronger reason to remove this. [#21460 (comment)](https://github.com/pytorch/pytorch/pull/21460#issuecomment-499617538)
> 
> I could change it to a flag possibly. But this is from a BC change five years ago haha I think it's been baked in by this point?

So I followed up with Ed, he said if there's no behavioral change this should be okay.

mlazos committed:
Update on "Remove out of order warning"

Rationale: https://github.com/pytorch/pytorch/pull/21460#issuecomment-499546570

In addition to above the main motivation is we will recompile constantly if that counter is incremented on each step. Additionally, this imposes the constraint that the LRScheduler should be created before we torch.compile the optimizer. Without this patching, the order does not matter. We *could* change the counter to a tensor or a flag, but since this warning is 5 years old I thought it was best to just remove.




[ghstack-poisoned]

mlazos committed:
Update on "Remove out of order warning"

Rationale: https://github.com/pytorch/pytorch/pull/21460#issuecomment-499546570

In addition to above the main motivation is we will recompile constantly if that counter is incremented on each step. Additionally, this imposes the constraint that the LRScheduler should be created before we torch.compile the optimizer. Without this patching, the order does not matter. We *could* change the counter to a tensor or a flag, but since this warning is 5 years old I thought it was best to just remove.




[ghstack-poisoned]

mlazos committed:
Update on "Remove out of order warning"

Rationale: https://github.com/pytorch/pytorch/pull/21460#issuecomment-499546570

In addition to above the main motivation is we will recompile constantly if that counter is incremented on each step. Additionally, this imposes the constraint that the LRScheduler should be created before we torch.compile the optimizer. Without this patching, the order does not matter. We *could* change the counter to a tensor or a flag, but since this warning is 5 years old I thought it was best to just remove.




[ghstack-poisoned]

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]

mlazos commented:
@janeyx99 this one should be ready for review as well

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]

janeyx99 commented on torch/optim/lr_scheduler.py:
```diff
@@ -76,39 +76,31 @@ def __init__(self, optimizer, last_epoch=-1, verbose="deprecated"):
+                # we've already patched
+                return opt.step
+
+            def wrap_step(step_fn):
```
For my understanding, how come you don't use the `@wraps(func)` pattern anymore? Did you write the code this way to avoid needing weakrefs? (This way the refs used will be freed once wrap_step exits?)

mlazos committed:
Update on "Swap warning counter to flag in LRScheduler"

This was a counter previously, this should be a flag to indicate whether or not the optimizer step has been called.




[ghstack-poisoned]
